Fix typo (JAVA_INLCUDES -> JAVA_INCLUDES)
Build dynamic library as well as static
Add ssw_cpp.* to library targets
Respect CPPFLAGS
--- SSW/src/Makefile.orig	2023-04-21 21:31:07.000000000 +0200
+++ SSW/src/Makefile	2023-05-03 15:46:33.246791776 +0200
@@ -1,17 +1,18 @@
-CC = gcc
-CXX = g++
-CFLAGS += -Wall -pipe -O2
+CC ?= gcc
+CXX ?= g++
+#CFLAGS += -Wall -pipe -O2
 #CFLAGS += -Wall -pipe -g -fsanitize=address	# for debug
-CXXFLAGS := $(CFLAGS)
+#CXXFLAGS := $(CFLAGS)
 LOBJS = ssw.o
 LCPPOBJS = ssw_cpp.o
 PROG = ssw_test
-LIB = libssw.so
+SOVERSION = 0
+LIB = libssw.so.$(SOVERSION)
 EXAMPLE = example_c
 EXAMPLE_CPP = example_cpp
 JAVA_JAR = ssw.jar
 JAVA_LIB = libsswjni.so
-JAVA_INLCUDES = -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux
+JAVA_INCLUDES = -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/linux
 JAVA_OBJ = ssw/Aligner.class ssw/Alignment.class ssw/Example.class

 .PHONY: all default java clean
@@ -21,10 +22,12 @@
 java: java_home $(JAVA_JAR) $(JAVA_LIB)
 all: default java

-$(LIB): ssw.c ssw.h
-	$(CC) $(CFLAGS) -fPIC -shared -rdynamic -o $@ $<
+$(LIB): ssw.c ssw.h ssw_cpp.h ssw_cpp.cpp
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -fPIC -shared -rdynamic -Wl,-soname,$(LIB) -o $@ $^ $(LDFLAGS)
+	ln -sf $(LIB) libssw.so

-$(PROG): main.c kseq.h
+$(PROG): main.c kseq.h $(LIB)
+	$(CC) $(CPPFLAGS) $(CFLAGS) -o $@ main.c kseq.h -L. -lssw -lm -lz $(LDFLAGS)

 $(EXAMPLE): example.c

@@ -32,15 +35,15 @@
 $(PROG) $(EXAMPLE): $(LOBJS)
 	$(CC) -o $@ $(filter-out %.h,$^) $(CFLAGS) $(LDFLAGS) -lm -lz -march=armv8-a+fp+simd+crypto+crc
 else # x86(Intel)
-$(PROG) $(EXAMPLE): $(LOBJS)
-	$(CC) -o $@ $(filter-out %.h,$^) $(CFLAGS) $(LDFLAGS) -lm -lz
+$(EXAMPLE): $(LOBJS)
+	$(CC) -o $@ $(filter-out %.h,$^) $(CPPFLAGS) $(CFLAGS) -lm -lz $(LDFLAGS)
 endif

 $(EXAMPLE_CPP): example.cpp $(LOBJS) $(LCPPOBJS)
-	$(CXX) -o $@ $^ $(CXXFLAGS) $(LDFLAGS) -lm -lz
+	$(CXX) -o $@ $^ $(CPPFLAGS) $(CXXFLAGS) -lm -lz $(LDFLAGS)

 $(JAVA_LIB): sswjni.c ssw.c ssw.h
-	$(CC) $(CFLAGS) $(JAVA_INLCUDES) -fPIC -shared -rdynamic -o $@ $< ssw.c
+	$(CC) $(CPPFLAGS) $(CFLAGS) $(JAVA_INCLUDES) -fPIC -shared -rdynamic -o $@ $< ssw.c $(LDFLAGS)

 $(JAVA_JAR): $(JAVA_OBJ)
 	jar cvfe $@ ssw.Example $^
@@ -55,10 +58,10 @@
 endif

 ssw.o: ssw.c ssw.h
-	$(CC) -c -o $@ $< $(CFLAGS)
+	$(CC) -c -o $@ $< $(CPPFLAGS) $(CFLAGS)

 ssw_cpp.o: ssw_cpp.cpp ssw_cpp.h ssw.h
-	$(CXX) -c -o $@ $< $(CXXFLAGS)
+	$(CXX) -c -o $@ $< $(CPPFLAGS) $(CXXFLAGS)

 clean:
 	-rm -f $(LOBJS) $(LCPPOBJS) $(PROG) $(LIB) $(EXAMPLE) $(EXAMPLE_CPP) $(JAVA_LIB) $(JAVA_JAR) $(JAVA_OBJ) *~
